### PR TITLE
docs: Add blog post on how we test vector

### PR DESCRIPTION
**Pull request by lukesteensen**
Eager for feedback on this, so don't hold back!

Running `make generate` locally seemed to choke on this for some reason (/cc @binarylogic), so this is just the raw markdown for now.

### CODE REVIEW

**Comment by binarylogic** [comment]
Thanks! I'll take a look and will review.

File: website/blog/2019-12-12-how-we-test-vector.md [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+to environment issues, less dependent on timing, etc, and they fulfill the
+purpose of giving us confidence that really does communicate with these external
+systems correctly.
+
+## Generative testing
+
+When we talked about unit tests, we mentioned that one area where they start to
+break down is when testing functions with a very large space of possible inputs.
+While you can do your best to think about that input space and come up with the
+edge cases that need handled, the problem is that you'll likely be thinking
```
Comment (Jeffail) [inline website/blog/2019-12-12-how-we-test-vector.md]: that need [to be] handled

File: website/blog/2019-12-12-how-we-test-vector.md [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+Closely related to property-based testing is the idea of fuzz testing. Fuzz
+testing is, at its most simplistic, just feeding your program random data and
+seeing if it breaks. There is a huge variety of tools and techniques for
+accomplishing this, many of which are remarkably sophisticated.
+
+One of the most popular and influential fuzz testing tools is [american fuzzy
+lop][7]. It is a coverage-guided, genetic fuzzer, which means that it watches
+your program's execution with various random inputs and uses that information to
+evolve new inputs that cover as many execution paths as possible. This makes it
+extremely effective at find edge cases where your program could crash, hang,
```
Comment (Jeffail) [inline website/blog/2019-12-12-how-we-test-vector.md]: at find[ing] edge cases

File: website/blog/2019-12-12-how-we-test-vector.md [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+than that fact that it will not panic and crash the program.
+
+Because of the way that a fuzzer instruments your program to track code coverage
+and observe crashes, it's generally not something used in the same way as unit
+tests you run on every build. There is a bit of setup required to expose
+a target that directly accepts raw bytes, build it with the proper
+instrumentation, and wrap it to be driven by the fuzzer. Since we are not
+actively changing the tokenizer parser, we simply did a session of fuzzing
+manually on a developer's machine (using the excellent [`cargo-fuzz`][13]),
+collected failing inputs and addresses them one by one, continuing until the
```
Comment (Jeffail) [inline website/blog/2019-12-12-how-we-test-vector.md]: and addresse~s~[d] them one by one

**Comment by binarylogic** [comment]
Nice work. The language and writing are excellent. A few changes I'd like to see to finish this off:

1. What is the user gaining by reading this post? I think it's learning about testing techniques and when to apply them. We can make this more clear in the intro and with a closing statement.
2. I'd like to think about the factors each test type contributes to. For example, reliability, effort, maintenance cost, bug discovery, and so on. It'd be interesting to compare these types across these factors so users know when to apply them (such as a table/matrix).
3. To extend the point above, I'd like to create a graphic that summarizes the testing strategies. We could do something as simple as a chart, where the y-axis is value and the x-axis is effort. I'm open to better ideas here, maybe something that better demonstrates the factors?
4. Finally, it might help, also, to include simple examples of each test.

**Comment by a-rodin** [comment]
> Running `make generate` locally seemed to choke on this for some reason (/cc @binarylogic), so this is just the raw markdown for now.

It is possible to run it using Docker:

```
./scripts/docker-run.sh checker make generate
```

That's the usual way I run it.

**Comment by binarylogic** [comment]
Yes, `make generate` also default to running through Docker now, so it should "just work".

File: website/blog/2020-02-12-how-we-test-vector.md [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+For other types of components, so much of the functionality is focused on
+communication with external systems that it can be difficult to isolate logic
+enough for a unit test. As much as possible, we try to critically assess that
+difficulty and use it to help find ways we can refactor the design to be more
+modular and therefore amenable to unit testing. Unit tests are an excellent
+source of design feedback, so we want to feel that pain and refactor when
+they're difficult to write.
+
+That being said, there are two specific places we often run into the limitations
+of unit tests. The first and more obvious is around pieces of code that are
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply website/blog/2020-02-12-how-we-test-vector.md]: I often prefer to embolden things. :) Eg `**fundamentally not isolated**` and `**size of the input space** and **number of potential paths**`. But both work great. Having am emphasis on them will help them stand out.

File: website/blog/2020-02-12-how-we-test-vector.md [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+done all we can to isolate logic into small, unit-testable functions, we still
+need to ensure the component as a whole does what it's supposed to. If unit
+tests tell you that something works in theory, integration tests tell you that
+it _should_ work in practice.
+
+Similar to unit tests, there are two major downsides to integration tests. The
+first is the same problem unit tests have with exhaustiveness, but intensely
+magnified. Because they often cover the interaction of multiple whole systems,
+integration tests will virtually never cover the combinatorial explosion of
+potential execution paths. Second, integration tests are often simply a pain to
```
Comment (Hoverbear) [inline website/blog/2020-02-12-how-we-test-vector.md]:
```suggestion
integration tests will virtually never cover all possible execution paths. Second, integration tests are often simply a pain to
```

File: website/blog/2020-02-12-how-we-test-vector.md [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+write and maintain. Given their dependence on external systems, they can be
+tedious to write, slow to run, and often flaky if their environment isn't set up
+just so.
+
+Takeaways:
+
+* While they can be a pain, integration tests are a valuable sanity check that
+  your system will actually work for users.
+* Don't overuse them trying to cover every possible scenario, because you can't.
```
Comment (Hoverbear) [inline website/blog/2020-02-12-how-we-test-vector.md]: I suggest adding

```markdown
* Integration tests may rely on external services or components.
```

File: website/blog/2020-02-12-how-we-test-vector.md [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+example-based strategies like unit and integration testing. Humans are bad at
+visualizing very large state spaces, and the state space of your code's possible
+inputs and execution paths is enormous. You also bring all the same biases when
+writing tests as you did when writing the original code.
+
+Generative tests address these issues by taking the human out of the equation
+and having the computer generate millions of example inputs. The tradeoff is
+that since you aren't simply hardcoding a list of inputs and their expected
+outputs anymore, you're forced to come up with more creative ways of identifying
+failures.
```
Comment (Hoverbear) [inline website/blog/2020-02-12-how-we-test-vector.md]:
```suggestion
outputs anymore, you're often forced to come up with creative ways of identifying
failures.
```

File: website/blog/2020-02-12-how-we-test-vector.md [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+
+* Taking time to "zoom out" and test your system as a user would can help
+  uncover blind spots and sanity-check behavior.
+* Evaluating similar tools can help build a better understanding of user
+  expectations.
+
+### Reliability tests
+
+A third category that we're currently working to integrate into
+`vector-test-harness` is something we're calling reliability tests. These are
```
Comment (Hoverbear) [inline website/blog/2020-02-12-how-we-test-vector.md]: Link to this!

```suggestion
[`vector-test-harness`](https://github.com/timberio/vector-test-harness) is something we're calling reliability tests. These are
```

File: website/blog/2020-02-12-how-we-test-vector.md [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+the environment at the time of the failure that you stand a chance at
+understanding and reproducing it. This task itself is a great test for our
+internal observability, and any issue we can't reproduce is a sign that our
+logging and metrics data needs to be improved.
+
+Another issue with these tests is that the vast majority of the time, nothing
+particularly interesting is happening. Since we want to find bugs as quickly as
+possible, we can supplement the randomness of the environment by injecting
+various types of faults on our own. There are a variety of tools for this, such
+as [Toxiproxy][10] and [Namazu][11].
```
Comment (Hoverbear) [inline website/blog/2020-02-12-how-we-test-vector.md]: Might want to mention `tc` since it's on most Linux hosts already. :)

**Comment by binarylogic** [comment]
> My only real nit: It's a bit long and it's very ...text. Do we have any sort of content to break it up? Maybe a nice image? Or maybe some colorful diagrams? Or maybe a couple examples of basic tests? Eg for unit tests, show a rust unit test.

I plan to [jazzify it](https://www.youtube.com/watch?v=xuPSIbABYVU). I plan on:

1. Adding an image to the top (radar chart)
2. Big callout quotes (medium style)
3. Code examples.

Definitely open to other suggestions. We're just getting our feet on the ground with content.

**Comment by Hoverbear** [comment]
@lukesteensen assigning this to myself to clean up and publish soonish!

**Comment by Hoverbear** [comment]
@lukesteensen I think this slipped off your plate, can I help you get it merged?

**Comment by lukesteensen** [comment]
@Hoverbear I believe @binarylogic had some thing he was planning to add to jazz things up, so I had left this on the backburner. I'd be happy to pick it up and push it across the finish line if no one would like to add anything else.

**Comment by Hoverbear** [comment]
Let's check with @binarylogic ! :)
